src/drivers/PortFolder/KMDFTestTemplate/Device.c

```diff
@@ -0,0 +1,199 @@
+        // with the device object. We define this structure in the device.h
+        // header file. DeviceGetContext is an inline function generated by
+        // using the WDF_DECLARE_CONTEXT_TYPE_WITH_NAME macro in device.h.
+        // This function will do the type checking and return the device context.
+        // If you pass a wrong object handle it will return NULL and assert if
+        // run under framework verifier mode.
+        //
+        deviceContext = DeviceGetContext(device);
+
+        //
+        // Initialize the context.
+        //
+        deviceContext->PrivateDeviceData = 0;
+
+        //
+        // Create a device interface so that applications can find and talk
+        // to us.
+        //
+        status = WdfDeviceCreateDeviceInterface(
+            device,
+            &GUID_DEVINTERFACE_KMDFTestTemplate,
+            NULL // ReferenceString
+            );
+
+        if (NT_SUCCESS(status)) {
+            //
+            // Initialize the I/O Package and any Queues
+            //
+            status = KMDFTestTemplateQueueInitialize(device);
+        }
+    }
+
+    return status;
+}
+
+NTSTATUS
+KMDFTestTemplateEvtDevicePrepareHardware(
+    _In_ WDFDEVICE Device,
+    _In_ WDFCMRESLIST ResourceList,
+    _In_ WDFCMRESLIST ResourceListTranslated
+    )
+/*++
+
+Routine Description:
+
+    In this callback, the driver does whatever is necessary to make the
+    hardware ready to use.  In the case of a USB device, this involves
+    reading and selecting descriptors.
+
+Arguments:
+
+    Device - handle to a device
+
+Return Value:
+
+    NT status value
+
+--*/
+{
+    NTSTATUS status;
+    PDEVICE_CONTEXT pDeviceContext;
+    WDF_USB_DEVICE_CREATE_CONFIG createParams;
+    WDF_USB_DEVICE_SELECT_CONFIG_PARAMS configParams;
+
+    UNREFERENCED_PARAMETER(ResourceList);
+    UNREFERENCED_PARAMETER(ResourceListTranslated);
+
+    PAGED_CODE();
+
+    TraceEvents(TRACE_LEVEL_INFORMATION, TRACE_DRIVER, "%!FUNC! Entry");
+
+    status = STATUS_SUCCESS;
+    pDeviceContext = DeviceGetContext(Device);
+
+    //
+    // Create a USB device handle so that we can communicate with the
+    // underlying USB stack. The WDFUSBDEVICE handle is used to query,
+    // configure, and manage all aspects of the USB device.
+    // These aspects include device properties, bus properties,
+    // and I/O creation and synchronization. We only create the device the first time
+    // PrepareHardware is called. If the device is restarted by pnp manager
+    // for resource rebalance, we will use the same device handle but then select
+    // the interfaces again because the USB stack could reconfigure the device on
+    // restart.
+    //
+    if (pDeviceContext->UsbDevice == NULL) {
+
+        //
+        // Specifying a client contract version of 602 enables us to query for
+        // and use the new capabilities of the USB driver stack for Windows 8.

```

Comment:
Will this cause any issues if the driver to be tested was built for pre Win8 OS?